mempool/src/shared_mempool.rs

```diff
@@ -259,6 +269,158 @@ async fn process_incoming_transactions<V>(
+            .send(Ok(response_data))
+            .map_err(|_| format_err!("[shared mempool] handling inbound RPC call timed out"))
+        {
+            error!(
+                "[shared mempool] failed to process batched transaction request, error: {:?}",
+                err
+            );
+        }
+    } else {
+        error!(
+            "[shared mempool] Did not get a response msg back from submit transactions from upstream peer / peer validator"
+        );
+    }
+}
+
+/// Tries to add transactions to local mempool
+/// returns RPC response that can be sent to callback
+/// peer_id is the PeerId of the node that sent the request
+async fn submit_transactions_to_mempool<V>(
+    smp: SharedMempool<V>,
+    peer_id: PeerId,
+    request: SubmitTransactionsRequest,
+) -> failure::Result<SubmitTransactionsResponse>
+where
+    V: TransactionValidation,
+{
+    /////////////////////////////////////////////
+    // convert from proto to SignedTransaction //
+    /////////////////////////////////////////////
+    let transactions: Vec<_> = request
+        .transactions
+        .clone()
+        .into_iter()
+        .filter_map(|txn| match SignedTransaction::try_from(txn.clone()) {
+            Ok(t) => Some(t),
+            Err(e) => {
+                // TODO make RPC response for this invalid transaction
+                // log error
+                security_log(SecurityEvent::InvalidTransactionMP)
+                    .error(&e)
+                    .data(&txn)
+                    .log();
+                None
+            }
+        })
+        .collect();
+
+    //////////////////////////////////
+    // validate transactions via VM //
+    //////////////////////////////////
+
+    // eagerly filter out transactions that were already committed
+    let account_states = join_all(
+        transactions
+            .iter()
+            .map(|t| get_account_state(smp.storage_read_client.clone(), t.sender())),
+    )
+    .await;
+
+    let transactions: Vec<_> = transactions
+        .into_iter()
+        .enumerate()
+        .filter_map(|(idx, t)| {
+            if let Ok((sequence_number, balance)) = account_states[idx] {
+                if t.sequence_number() >= sequence_number {
+                    return Some((t, sequence_number, balance));
+                }
+            }
+            None
+        })
+        .collect();
+
+    let validations = join_all(
+        transactions
+            .iter()
+            .map(|t| smp.validator.validate_transaction(t.0.clone()).compat()),
+    )
+    .await;
+
+    ///////////////////////////////
+    // add txns to local mempool //
+    ///////////////////////////////
+
+    // The [`TimelineState`] for adding txns to mempool should only be TimelineState::NonQualified,
+    // (i.e. this node should never try to broadcast it) if this MempoolSyncMsg came from one validator
+    // to another validator in the same validator network.
+    let timeline_state = if smp
```

Comment:
good thinking. I wonder if we can do it another way without any locking. Similar to state synchronizer you can just analyze which network is used for delivering this message. I'd pass it just as argument to function call